lib/Sema/TypeCheckStorage.cpp

```diff
@@ -1421,12 +1421,25 @@ synthesizeObservedSetterBody(AccessorDecl *Set, TargetImpl target,
   SmallVector<ASTNode, 6> SetterBody;
 
   auto callObserver = [&](AccessorDecl *observer, VarDecl *arg) {
+    // If we're not loading the old value, then remove the oldValue parameter
+    // and re-compute the type.
+    if (!arg) {
+      observer->setParameters(ParameterList::createEmpty(Ctx));
+      // Recompute the interface type
+      observer->setInterfaceType(Type());
+      (void)observer->getInterfaceType();
+    }
```

Comment:
@slavapestov Would it be reasonable / more correct to rewrite the function parameter clause in the request?

Reply:
You can't clear cached values like that. We should add an assertion to setInterfaceType() that prevents an empty Type from being stored there, but I think CSDiag still does this.

Anyway, you need to refactor things so that the interface type can be computed from the body in one shot. Right now type checking the body depends on the interface type; I'm not sure how feasible it would be to reverse that ordering.

An alternative approach is to detect usages of oldValue in a purely syntactic manner, by checking for DeclRefExprs filled in by the parser's local name lookup. It won't be as accurate, but perhaps a post-pass can emit a warning if an UnresolvedDeclRefExpr later becomes a reference to oldValue, which would not be picked up by the interface type computation.

Reply:
Indeed https://github.com/apple/swift/pull/28948/files